Signal messenger attachments as bytes support

burnsie-la opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

This is my first ever PR for any open source project - I have no idea if I'm doing the right things, so please go easy on me if I've done something wrong! :)

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

This adds the ability for the Signal Messenger integration to support attachments downloaded from a URL/API. Previously this integration would only allow you to attach files read from disk.  This means you can attach images from the camera snapshot API or from Frigate's notifcation API to a Signal message

[Git diff](https://github.com/bbernhard/pysignalclirestapi/compare/6f148bb..0697626) between `0.3.4` and `0.3.17` of the [pysignalclirestapi package](https://github.com/bbernhard/pysignalclirestapi)

## Type of change
- [x] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [x] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

BREAKING CHANGE: We have removed the deprecated **Attachment** option, which accepted a single file path.  This was replaced by the **Attachments** option several months ago, which accepts a list of file paths for multiple attachment support.  Use of the **Attachment** option previously returned a warning indicating that this was deprecated.

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]
[Pull Request for Documentation Update](https://github.com/home-assistant/home-assistant.io/pull/20762)

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
I bumped the version of https://github.com/bbernhard/pysignalclirestapi from 0.3.4 to 0.3.17 to take advantage of the new attachment functionality provided by the API.

- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
- [x] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

I'm more of a scripter than a developer - it took me weeks to get this thing working! I don't even really know how to review PRs.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

burnsie-la committed:
Update manifest.json

Version bump pysignalclirestapi to 0.3.17 in order to support v2 signal cli API.

burnsie-la committed:
Update notify.py

Added support to request a url and attach it to a Signal message as bytes.

burnsie-la committed:
Added tests

burnsie-la committed:
Updated URL to URL_ATTACHMENT for clarity

ianByrne committed:
Update Signal tests

burnsie-la committed:
Updated requirements.txt files

ianByrne committed:
Merge branch 'signal_messenger_attachments_as_bytes_support' of https://github.com/burnsie-la/core into signal_messenger_attachments_as_bytes_support_testfix

ianByrne committed:
Remove whitespace

burnsie-la committed:
Merge pull request #3 from ianByrne/signal_messenger_attachments_as_bytes_support_testfix

Signal messenger attachments as bytes support testfix

homeassistant commented:
Hi @ianByrne, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;core#62311).

Once you do that we will be able to review and accept this pull request.

Thanks!

burnsie-la commented:
I created a pull request for the documentation as well: https://github.com/home-assistant/home-assistant.io/pull/20759

elupus commented on tests/components/signal_messenger/test_notify.py:
```diff
@@ -91,6 +92,27 @@ def send_message_with_attachment(signal_notification_service, deprecated=False):
+    assert signal_requests_mock.called
+    assert signal_requests_mock.call_count == 3
+    assert_sending_requests(signal_requests_mock, 1)
+
```
This year really aight to check the content of the data sent.

ianByrne replied:
Added content check

elupus commented on homeassistant/components/signal_messenger/manifest.json:
```diff
@@ -3,6 +3,6 @@
   "documentation": "https://www.home-assistant.io/integrations/signal_messenger",
   "codeowners": ["@bbernhard"],
-  "requirements": ["pysignalclirestapi==0.3.4"],
+  "requirements": ["pysignalclirestapi==0.3.17"],
```
Since this bumps a dependency. The pull request needs a link to the changelog between revisions.

ianByrne replied:
There doesn't appear to be a maintained changelog for this package (apologies if I'm misunderstanding your request):
https://github.com/bbernhard/pysignalclirestapi/releases
https://pypi.org/project/pysignalclirestapi/#history

elupus replied:
You can link to a diff between the commit of the releases.

ianByrne replied:
I've now added that as a comment to the PR

elupus commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -71,9 +72,21 @@ def send_message(self, message="", **kwargs):
+                        attachments_as_bytes.append(
+                            (requests.get(url, verify=False, timeout=10)).content
+                        )
+                    except Exception as ex:
```
Feel like there should be some size limitation here? This could take down the server if you ask it to send a very large file.

ianByrne replied:
Added size limits checks and tests (d8e59e9b32d64a881b57421a921f317dd169e049)

ianByrne committed:
Add content check to Signal test

ianByrne commented:
[Git diff](https://github.com/bbernhard/pysignalclirestapi/compare/6f148bb..0697626) between `0.3.4` and `0.3.17` of the [pysignalclirestapi package](https://github.com/bbernhard/pysignalclirestapi)

burnsie-la committed:
Added size limits to request

burnsie-la committed:
Added max allowed size of attachment

ianByrne committed:
Add tests for signal attachments

ianByrne commented:
@elupus - @burnsie-la  and I have now implemented your suggestions. This is our first PR - do you know if there is anything else needed on our end? Or what the next steps are? Many thanks

elupus commented:
Needs review by maintainer or somebody who know the integration.

bbernhard commented:
Thanks for the extension of the Signal Messenger integration @burnsie-la and @ianByrne.

I haven't contributed anything in a while to Home Assistant (I am still actively maintaining the signal-cli-rest-api docker container though, but haven't had time to work on the `signal_messenger` HA integration), but to me the PR looks good and I would be happy if it could be merged soon!

ianByrne committed:
Refactor for pylint

ianByrne commented:
@elupus I've now pushed an update to address the Pylint issues. Would it be possible to trigger the workflow again?

Also, would you know if there's an easy way to run all the tests locally so that I can catch these issues sooner? The automated tests that occur on local when running the commit hooks didn't seem to pick this one up.

ianByrne committed:
Add tests

ianByrne committed:
Merge pull request #4 from burnsie-la/signal_fixes

Add tests

ianByrne commented:
Sorry @elupus but would it be possible to trigger the workflows once more 😶 I've now added extra tests for more code coverage, hopefully that is the last thing 🙄 many thanks

ianByrne committed:
Add coverage for SignalCliRestApiError

ianByrne commented:
The CI codecov checks are failing, however after [discussions in Discord](https://discordapp.com/channels/330944238910963714/330990195199442944/925876094811385866) it looks like this is not due to the changes made in this PR (the touched code does indeed have 100% code coverage):

![image](https://user-images.githubusercontent.com/1918098/147708171-4328ebb1-df06-4e38-94e8-45a888f736dd.png)

So this PR is ready for review and merge 🙂

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -50,30 +53,78 @@ def __init__(self, recp_nrs, signal_cli_rest_api):
+        if ATTR_FILENAMES in data:
+            filenames = data[ATTR_FILENAMES]
+        if ATTR_FILENAME in data:
+            _LOGGER.warning(
```
This option ought to be removed as 0.108 has long been passed

ianByrne replied:
Removed in d9415bcc03823e892c547210f4faf4435934fbfc

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -50,30 +53,78 @@ def __init__(self, recp_nrs, signal_cli_rest_api):
+        if data is None:
+            return None
+        if ATTR_FILENAMES in data:
+            filenames = data[ATTR_FILENAMES]
```
Can we verify that `data[ATTR_FILENAMES]` is in fact a list and either throw an error or silently handle it if it isn't?

ianByrne replied:
Type check added in d9415bcc03823e892c547210f4faf4435934fbfc

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -50,30 +53,78 @@ def __init__(self, recp_nrs, signal_cli_rest_api):
+        if data is None or ATTR_URLS not in data:
+            return attachments_as_bytes
+
+        urls = data[ATTR_URLS]
```
Can we also verify that this is a list?

ianByrne replied:
Type check added in d9415bcc03823e892c547210f4faf4435934fbfc

alexyao2015 commented:
General note: Can typing be added?

ianByrne committed:
Add typing, handle type errors

ianByrne committed:
Fix header type

ianByrne committed:
Fix pylint errors

ianByrne committed:
Merge pull request #5 from burnsie-la/typing

Typing

ianByrne commented:
> General note: Can typing be added?

Thanks @alexyao2015 - I've made changes in d9415bcc03823e892c547210f4faf4435934fbfc to address your suggestions

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -43,37 +54,88 @@ def get_service(hass, config, discovery_info=None):
             raise ex
+
+    @staticmethod
+    def get_filenames(data: Any) -> list[str] | None:
```
It seems that HA is still running Python 3.9 and as such, this syntax has only [become supported with 3.10](https://www.python.org/dev/peps/pep-0604/). You will need to revert back to the `typing.Optional` syntax.

```
Python 3.9.9 (main, Nov 17 2021, 16:20:40) 
[GCC 10.2.1 20210110] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> def dummy() -> list | None:
...     pass
... 
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -43,37 +54,88 @@ def get_service(hass, config, discovery_info=None):
+    @staticmethod
+    def get_attachments_as_bytes(
+        data: Any, attachment_size_limit: int
+    ) -> list[bytearray] | None:
```
Same here.

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -1,23 +1,30 @@
 CONF_RECP_NR = "recipients"
 CONF_SIGNAL_CLI_REST_API = "url"
-ATTR_FILENAME = "attachment"
+CONF_MAX_ALLOWED_DOWNLOAD_SIZE_BYTES = 52428800
```
Add this config to the `PLATFORM_SCHEMA` with this as the default? Also would need to document it.

ianByrne replied:
I think the name of this const is a bit misleading - it's not really a config/PLATFORM_SCHEMA property. I'll update the name to `ATTACHMENT_SIZE_LIMIT` in the next commit.

@burnsie-la can you please update your [docs PR](https://github.com/home-assistant/home-assistant.io/pull/20762) to mention the limit

alexyao2015 replied:
[Seems that the limit is 100mb](https://www.reddit.com/r/signal/comments/astgm2/comment/egxaq6i/). From my testing it seems that it is 100mb per file, not per message. I was able to upload files totalling over 100mb in a single message. Can this be enforced instead of the 50mb limit? Additionally can it be enforced on local files? Perhaps also just pass the local files as bytes to the lib along with url attachments.

burnsie-la replied:
The 100MB limit that you linked to is the Signal service limit. We added this limitation to prevent causing issues with the HA server and/or Signal Docker container in the event that you inadvertently attempt to attach a massive file.

50MB was chosen to keep the attachment limits consistent with the [Home Assistant companion app](https://companion.home-assistant.io/docs/notifications/notification-attachments#supported-media-types).

burnsie-la replied:
This PR is to add support for Attachments as Bytes from URLs.  We can revisit the local file handling in a future PR.

alexyao2015 replied:
I see but this isn't the companion app. It's a separate service with its own limits.

alexyao2015 commented:
Its looking great! Thanks for all the hard work!

Additionally, if you can `Resolve conversation` on the reviews that have been accounted for, that would help to condense the conversation.

alexyao2015 commented on homeassistant/components/signal_messenger/manifest.json:
```diff
@@ -3,6 +3,6 @@
   "documentation": "https://www.home-assistant.io/integrations/signal_messenger",
   "codeowners": ["@bbernhard"],
-  "requirements": ["pysignalclirestapi==0.3.4"],
+  "requirements": ["pysignalclirestapi==0.3.17"],
```
```suggestion
  "requirements": ["pysignalclirestapi==0.3.17", "future==0.18.2"],
```
Seems that the `future` package is also required due to the Python 2.7 backport with this particular package. The lib  itself does not define its own requirements.

ianByrne replied:
Ah I was wondering why I had to install that library separately! I had similar issues with `sqlalchemy` and `zeroconf` - do you know if those need to be added here too? Or were those missing packages something unrelated?

alexyao2015 replied:
If that is with HA, I haven't encountered those issues. If you are talking generally, most packages define `REQUIRES` so pip can automatically install dependencies. This particular package does which causes this issue.

alexyao2015 replied:
It seems that this issue has been resolved by the library with the newest version. Suggest bumping the library and removing the future dependency. https://github.com/bbernhard/pysignalclirestapi/releases/tag/0.3.18

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -43,37 +54,88 @@ def get_service(hass, config, discovery_info=None):
+        urls = data[ATTR_URLS]
+        for url in urls:
+            try:
+                resp = requests.get(url, verify=False, timeout=10, stream=True)
```
Why is verify set to off? I see some people may want it off, but it can pose a security risk for some. Can this instead be a config option with it defaulted to on?

ianByrne replied:
That's a good idea - the particular use case we had was downloading files from a camera's local API, so the connection was not https. Making this optional makes more good sense.

alexyao2015 replied:
If it's a local camera, it should be using http? Turning verify off only disables ssl certificate verification.

ianByrne replied:
This is now made optional in a25bbc70353e0aad07b3e769e58794e6d682748a. Docs have been updated to show the new attribute

ianByrne commented:
> Additionally, if you can `Resolve conversation` on the reviews that have been accounted for, that would help to condense the conversation.

I'm a GitHub noob and was puzzled that there didn't seem to be a `resolve` option anywhere - after some digging around, turns out this is likely because I'm not the PR creator or someone with Write access. I'll work with @burnsie-la tomorrow to close out the old ones, thanks for the tip!

alexyao2015 commented:
Perhaps a config flow can also be added to this to replace the old yaml config as well.

burnsie-la commented:
> Perhaps a config flow can also be added to this to replace the old yaml config as well.

That is a great suggestion, but out of scope for what we're trying to achieve with this PR.

ianByrne committed:
Update homeassistant/components/signal_messenger/manifest.json

Co-authored-by: Alex <33379584+alexyao2015@users.noreply.github.com>

ianByrne committed:
Make verifyUrls optional, bump py package

ianByrne committed:
Rename attr to verifySsl

burnsie-la commented:
I updated this PR to indicated that it is now a Breaking Change as we removed the deprecated Attachment option.  Thanks @alexyao2015 for the suggestion to improve the integration!

frenck commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -43,37 +55,93 @@ def get_service(hass, config, discovery_info=None):
+
+        return None
+
+    @staticmethod
+    def get_attachments_as_bytes(
+        data: Any, attachment_size_limit: int
+    ) -> list[bytearray] | None:
```
IMHO this is a no-go.
It downloads files from a possible random URL and sends it out.

burnsie-la replied:
> IMHO this is a no-go. It downloads files from a possible random URL and sends it out as attachment.

What are your specific concerns with this approach? If we were to only support images, videos and audio would that be acceptable?

frenck replied:
My concern is that using a call a random file can be downloaded and shipped off our platform. That sounds not acceptable to me.

burnsie-la replied:
The HA Companion app shows an example of loading an attachment from Github - is this different in some way?

https://companion.home-assistant.io/docs/notifications/notification-attachments#example-service-call

frenck replied:
Yes, as those are not sent over as binary data.
In that case, the app receiving is responsible for handling external data responsibly.

burnsie-la replied:
I would love any advice you or anyone else may have on how we can achieve our goals of attaching files from an API to the message in a safe and acceptable way. For this use case I am looking to attach images and videos from the [Frigate Notification API](https://docs.frigate.video/integrations/home-assistant#notification-api) and sending them to a Signal message.

It appears that the Telegram integration does this the same way as we have designed:
https://www.home-assistant.io/integrations/telegram/#photo-support
https://github.com/home-assistant/core/blob/dev/homeassistant/components/telegram_bot/__init__.py#L257

alexyao2015 replied:
I agree it does seem like a security risk, but in reality, home assistant is a hub and is going to connect to external sources. Sure, the code to download the data could be moved to the library (as in the mobile app), but it's still home assistant that is running the code from the library. 

An alternative is to limit this to the home assistant external/internal URL's or provide a whitelist of sorts, however that would greatly limit functionality. Not to mention that as @burnsie-la mentioned the Telegram integration already does this, and I would not be surprised if other components do this as well. Perhaps there would need to be a greater architecture change around home assistant as a whole to address @frenck's comment.

alexyao2015 replied:
@burnsie-la It seems like the best way to protect the user and the platform would be to implement it like the Slack integration with a URL whitelist defined in the home assistant core config. 

See below:
https://www.home-assistant.io/integrations/slack/

https://www.home-assistant.io/docs/configuration/basic/#allowlist_external_urls

frenck replied:
@burnsie-la The above solution/suggestion by @alexyao2015 is indeed the right solution to this, which can/would be accepted.

ianByrne replied:
Thanks - now added in 9d6d7b1

alexyao2015 commented on homeassistant/components/signal_messenger/notify.py:
```diff
@@ -43,37 +55,93 @@ def get_service(hass, config, discovery_info=None):
+            return None
+
+        if ATTR_FILENAMES in data:
+            if isinstance(data[ATTR_FILENAMES], list):
```
Validate date using voluptuous?

See here for an example: https://github.com/home-assistant/core/blob/56c577c8321d543197c469eceaaefe660ed2b698/homeassistant/components/slack/notify.py#L44

ianByrne replied:
Thanks - now added in 9d6d7b1

ianByrne committed:
Add vol validation

ianByrne committed:
Implement allowlist_external_urls

ianByrne committed:
Log attachment size errors

ianByrne committed:
Merge pull request #6 from burnsie-la/updates

Updates

frenck reviewed (approved):
Looks great, thanks, @burnsie-la 👍
